### PR TITLE
Added possibility to set a general admin_code

### PR DESCRIPTION
**Pull request by Nexotap**

### CODE REVIEW

**Comment by OskarStark** [comment]
ping @rande

**Comment by greg0ire** [comment]
Please rebase on the latest commit, it will fix the build.

**Comment by Nexotap** [comment]
Bear with me, this is my first PR. It already took a while for me to get this far. Can you please help me how to rebase? And should I change it to defaultAdminCode?

**Comment by greg0ire** [comment]
No problem, I'll help you :)

First thing you should know : there is no risk of loosing your work. Once you add a change in git, you can recover it (see for instance, `git reflog`). With that in mind, let's rebase!

1. Run `git fetch` to get the latest commits from the main repo. If you look at the output of `git log --graph --decorate --all`, you should see master and origin/master are no longer the same.
2. Run `git rebase origin/master
3. Make the change to `defaultAdminCode`, I really think "default" is better than "general", because it conveys the meaning that it is taken into account only if an admin code is not specified
4. Run `git commit --amend`
5. Force push your work here : `git push --force`

Report here if you fail at any step.

**Comment by Nexotap** [comment]
I really appreciate your help. Alltough I'm a bit confused. I'm using GitDesktop for this. I guess, I synced somehow, and in a way my changes are commited. But I have no clue what goes on behind the scene.
I've read about forking here: https://help.github.com/articles/fork-a-repo/
There I see a different upstream than my repo. But if I look at my repo within GitShell, they are the same.
Would you recommend to use the shell instead of the desktop app?

**Comment by greg0ire** [comment]
Woops didn't see your answer. I totally recommend using the shell though, because this is what I use, so I can help you better that way, and you will get a better understanding of what happens behind the scenes when you use GitDesktop.

**Comment by greg0ire** [comment]
What you did with GitDesktop does not look good though, because you now have 3 commits, when there should be only one (we don't care about the name change, do we?).

**Comment by greg0ire** [comment]
I suggest you do step 1. and 2. and then, tell us what `git log --graph --decorate` looks like

**Comment by greg0ire** [comment]
Your first goal is to get rid of the "Merge remote-tracking branch 'refs/remotes/sonata-project/master" commit first. The rebase should be able to do that automatically : it should see that this commit is useless and throw it away.

**Comment by Nexotap** [comment]
It seems, only 2 commits left. But no luck to have only 1.

git remote -v tells me:

origin  https://github.com/Nexotap/SonataAdminBundle.git (fetch)
origin  https://github.com/Nexotap/SonataAdminBundle.git (push)
sonata-project  https://github.com/sonata-project/SonataAdminBundle.git (fetch)
sonata-project  https://github.com/sonata-project/SonataAdminBundle.git (push)

Is this how it should be?


The beginning of git log --graph --decorate looks like this:

* commit 1e7e999d7e9e19dd892fe6af41620cb95d028cca (HEAD -> master, origin/master, origin/HEAD)
| Author: Nexotap <nexotap@labrat.ch>
| Date:   Fri Jan 29 08:41:53 2016 +0100
|
|     Change the variable from adminCode to defaultAdminCode
|
|     Accordingly to the  variable, also changed the functionname.
|
|     Added possibility to set a default admin_code
|
* commit 09607749a4f6f448478eb3da1e4d33d1baed835b
| Author: Nexotap <nexotap@labrat.ch>
| Date:   Wed Jan 27 13:44:39 2016 +0100
|
|     Added possibility to set a general admin_code
|
*   commit 3294a3aec754ef5d1a16e9db292eae2e4a509c10 (sonata-project/master)
|\  Merge: 750b269 425c16a
| | Author: Oskar Stark <oskarstark@googlemail.com>
| | Date:   Sat Jan 30 01:00:23 2016 +0100
| |
| |     Merge pull request #3559 from core23/action-buttons
| |
| |     Customize action buttons in admin class
| |
| * commit 425c16a892e2fdea6b470fd0dea0eed71153d90b
| | Author: core23 <mail@core23.de>
| | Date:   Mon Jan 25 19:05:02 2016 +0100
| |
| |     Customize action buttons in admin class
| |
* |   commit 750b26988c90a1004710a3d950a4caa755d55b39


I tried forward and backward, but still I'm helpless confused.
What about the failed checks. Do I have to worry about them?

**Comment by greg0ire** [comment]
Ok, so it is starting to look very good. We'll worry about the failed checks, but later. Now how to get one commit instead of 2 ? 

1. Run `git rebase --interactive sonata-project/master`. You will be presented with an editor showing your two commits (all the commits since sonata-project/master). At the beginning of each line, there will be an action.
2. Go to the second line and change `pick` to `fixup`, and quit your editor. You should be left with one commit only
3. Force push again, with `git push --force`

**Comment by Nexotap** [comment]
That's like a birth. But it worked. If I get it correct, the fixup gets ride of the other commits?

**Comment by greg0ire** [comment]
The fixup fixes up the commit with the commit above. In the editor, here is what you can read : 

> #  s, squash = use commit, but meld into previous commit
#  f, fixup = like "squash", but discard this commit's log message

**Comment by greg0ire** [comment]
Ok so now, let's have a look at the tests. What fails here is `make cs` : 

> 1) Form/FormMapper.php (phpdoc_short_description)

File: Form/FormMapper.php [hunk 1]
```diff
@@ -51,6 +56,16 @@ public function reorder(array $keys)
     }
 
     /**
+     * Sets the default admin_code
+     *
+     * @param $defaultAdminCode
```
Comment (greg0ire) [inline Form/FormMapper.php]: Should probably be something like `@param string $defaultAdminCode`

File: Form/FormMapper.php [hunk 2]
```diff
@@ -64,6 +79,11 @@ public function add($name, $type = null, array $options = array(), array $fieldD
             return $this;
         }
 
+        // If a default  admin_code exist, and none is given, set the default admin_code
```
Comment (greg0ire) [inline Form/FormMapper.php]: There is an extra space after "default"

**Comment by greg0ire** [comment]
Try to fix the 2 aforementioned problems, then run `git add --patch` and `git commit --amend`. Amend is a lot like fixup, but you do not create a commit you fixup with the previous one. Instead, you immediately modify the previous commit. Then again, use `git push --force`.

**Comment by greg0ire** [comment]
I don't know if the problem with the short description is the missing type or just the missing description.

**Comment by greg0ire** [comment]
Still not good, and the message has slightly changed : 

> 1) Form/FormMapper.php (phpdoc_short_description, phpdoc_var_without_name)

Can you try running `make cs` ?

**Comment by Nexotap** [comment]
I have no make command. Working on windows here.

**Comment by greg0ire** [comment]
Have a look at the `Makefile` then, and run whatever is under `cs` ;)

**Comment by greg0ire** [comment]
If you don't have the underlying command, you can follow these instructions : http://cs.sensiolabs.org/

**Comment by Nexotap** [comment]
Ran that command. Another new thing :astonished: 
I now have lots of warnings about LF beeing replaced with CRLF.  The file will have its original line endings in my working directory.
But I think, its getting better :)

**Comment by greg0ire** [comment]
Oh I see. Make sure you configure your editor to have UNIX line-endings, and you should avoid this kind of thing forever :)

**Comment by greg0ire** [comment]
I think it is indeed a lot better! Now, how about tests and docs?

**Comment by greg0ire** [comment]
Or maybe you want to wait to see whether you PR might be accepted first? @core23 @OskarStark ? Any comments on this PR?

**Comment by Nexotap** [comment]
@greg0ire 
Thank you very much. I appreciate your help. :smile:
Tests and docs? Awww. I think I'll let the newly gained knowledge settle down a bit and see if my first PR gets accepted as it is :wink: 

While beeing here, might I ask 2 offtopic questions?

When developing on something like this. Do you get symfony with composer and within the vendor directory you checkout the fork?

What is your work environment? I had windows and wamp, but the apache seems to be slow. then I've tried it with vagrant. Not to bad, but I had to remap the cache and vendor library inside the vagrant box. Symfony was too slow. But then i have the problem, things are cluttered on 2 different systems. My newest idea is a complete Ubuntu running as a virtual machine, and doing everything there, with a shared folder from windows where the sources are. The botleneck now is the mounted folders as vboxfs. I'm quite desperate in this matter.

**Comment by greg0ire** [comment]
I usually already have a project with the vendor in it, but doing as you do is a good way to go :) Alternatively, you can specify `dev-master` for the admin bundle in your project and let Composer do the checkout for you.

Regarding the work environment, I use docker with [this set of images I built](https://github.com/greg0ire/dockerony). Instead of apache, it uses nginx + php-fpm

I run docker inside Ubuntu, so that I don't have the bottleneck you speak about. Why do you need Windows at all exactly ?

**Comment by Nexotap** [comment]
Good question about windows. I also use Products like Photoshop and Illustrator alot. Unhappily they dont exists on Linux. Tried Gimp several times, but I can't really make friends with it. I also had once Dualboot, but it was sure, as soon as I booted Linux, I had to do some gfx stuff ;) But maybe I should overthink my workflow?

**Comment by greg0ire** [comment]
Maybe you could use a Windows VM inside linux rather than the other way around? Or, if you work as a team, have your admin sys setup a remote Windows VM with Photoshop / Illustrator on it for use by all the team?

**Comment by Nexotap** [comment]
Well, I work alone, so no admin. I used to try Vine, a long time ago, but it wasnt satisfied. Never thought of running Windows as VM inside Linux. Will try that. Thank you very much.

Hopefully, next time I wont be such a pain.

**Comment by greg0ire** [comment]
> Hopefully, next time I wont be such a pain.

At least you managed to do it, thank your for not giving up!

**Comment by core23** [comment]
Not sure if we need some tests for this

**Comment by greg0ire** [comment]
I'm currently trying to write tests for this, but I'm not sure it makes sense, correct me if I'm wrong @Nexotap : if I set a default admin code to `admin.car`, every thing I add to the form mapper will have `admin.car` as `admin_code` (unless I specify something else of course, but that would be tedious).

**Comment by Nexotap** [comment]
That's exactly how it's supposed to be. The reason which brought me to this solution, is quite simple. I was in need of 2 different admin classes, which shared the same entity. I didnt wanted to write an admin code to every single form item.

As an example: Asuming we have 1 partner class, whereas a partner can be a supplier or a client. PersonallyI would have used 1 admin class and used a filter for that matter. But as clients sometimes can be a bit complicated ;) 
They wanted to have supplier and clients separated. How to explain that they share the same table with just a flag if its a supplier or client or both?
My solution, 2 admin classes.

Maybe my approach was wrong and the whole thing isnt necessary?

**Comment by greg0ire** [comment]
I don't understand, but maybe you could show us your code?

**Comment by Nexotap** [comment]
You can take a look at those files here:
https://github.com/Nexotap/Ivory/blob/master/src/AppBundle/Admin/PartnerAdmin.php
https://github.com/Nexotap/Ivory/blob/master/src/AppBundle/Admin/CustomerAdmin.php
https://github.com/Nexotap/Ivory/blob/master/src/AppBundle/Admin/SupplierAdmin.php

Those 3 Classes share the same Entity:
https://github.com/Nexotap/Ivory/blob/master/src/AppBundle/Entity/Partner.php

The admin config can be found here:
https://github.com/Nexotap/Ivory/blob/master/app/config/admin.yml

Please don't judge me by the code. It's a bit messy and not up to date. I hope that helps to understand what I mean.

**Comment by greg0ire** [comment]
I think the way to go would have been to use inheritance, which is supported by sonata admin. Not sure why you need the `admin_code` for fields that will be guessed as "text" fields though. Correct me if I'm wrong, but `admin_code` is used when you are referring to another entity, and this entity has several admin classes (it happened to me in my app, where I sometimes need to do things like this : `->add('library', null, [], ['admin_code' => 'admin.library']);`. Here `library` is an entity).

**Comment by Nexotap** [comment]
Now if I think about it once again, inheritance should do the work as well. Maybe I was too immersed in finding a solution that I wasn't looking at the easiest way which was directly in front of my eyes. :dizzy_face:
Not sure if this feature is still necessary then.

**Comment by greg0ire** [comment]
>  Not sure if this feature is still necessary then.

I'm afraid it isn't… sorry :(

Can you close this? Maybe someone will reopen it later if it turns out to be useful after all…

**Comment by Nexotap** [comment]
I'll get over it ;) At least I learned something about pull requests and such. Hopefully I wasn't too much of a pain.

**Comment by greg0ire** [comment]
> I'll get over it ;) At least I learned something about pull requests and such. Hopefully I wasn't too much of a pain.

Well the pain is greatly alleviated when people manage to see the positive side of the experience like you do! Glad you learnt something :)

**Comment by clytemnestra** [comment]
@Nexotap @greg0ire How would inheritance solve this problem? I'm doing the same thing nexotap did with the admin classes and getting that error. What is the solution to the problem? You still need to define the 2 admin classes as services and specify the entity for them, which throws the error.

**Comment by greg0ire** [comment]
Inheritance can improve @Nexotap 's code. There was no real solution here, and I think this might in fact be relevant : someone in my team had the problem when adding a new Admin for an entity afterwards. If you want, you can pick up this PR where @Nexotap left it.

**Comment by clytemnestra** [comment]
Well, I found a temporary solution by doing this https://github.com/sonata-project/SonataAdminBundle/issues/2212#issuecomment-223255198

Not sure for what that commented check is, but it's much cleaner than adding admin_code everywhere.

**Comment by greg0ire** [comment]
> much cleaner

Well if you comment this, how are you going to be sure the intended admin is in use?

**Comment by clytemnestra** [comment]
Well if multiple admin classes use the same entity, doesn't it make sense to use it only once if it appears multiple times? That's what $this->adminClasses[$class][0] does, since that array would have the same entity class multiple times, so it just takes the first value.

**Comment by greg0ire** [comment]
@Nexotap 's example turned out to be not so valid. Maybe yours is, can you show us?

**Comment by clytemnestra** [comment]
Something like so:

```yaml

services:
    med_app.admin.user_medic:
        class: Application\Sonata\AdminBundle\Admin\MedicsAdmin
        arguments: [~, MedAppBundle\Entity\User, ~]
        tags:
            - {name: sonata.admin, manager_type: orm, group: Users, label: Medics}
        calls:
             - [ setTemplate, [outer_list_rows_mosaic, MedAppBundle:AdminSonata/CRUD/User:list_outer_rows_mosaic.html.twig]]


    med_app.admin.user_pacient:
         class: Application\Sonata\AdminBundle\Admin\PatientsAdmin
         arguments: [~, MedAppBundle\Entity\User, ~]
         tags:
             - {name: sonata.admin, manager_type: orm, group: Users, label: Patients}
         calls:
              - [ setTemplate, [outer_list_rows_mosaic, MedAppBundle:AdminSonata/CRUD/User:list_outer_rows_mosaic.html.twig]]

```

```php
<?php

namespace Application\Sonata\AdminBundle\Admin;

use Doctrine\ORM\QueryBuilder;
use Sonata\AdminBundle\Admin\AbstractAdmin;
use Sonata\AdminBundle\Datagrid\DatagridMapper;
use Sonata\AdminBundle\Datagrid\ListMapper;
use Sonata\AdminBundle\Form\FormMapper;
use Sonata\AdminBundle\Route\RouteCollection;
use Sonata\AdminBundle\Show\ShowMapper;
use Symfony\Component\Form\Extension\Core\Type\TextType;

class MedicsAdmin extends AbstractAdmin
{

    protected $baseRoutePattern = 'medics';
    protected $baseRouteName = 'medics';
    protected $classnameLabel = 'Medics';


        protected $datagridValues = array(
            '_page' => 1,
            '_sort_order' => 'DESC',
            '_sort_by' => 'id',
        );



        public function getFilterParameters()
        {
            $this->datagridValues = array_merge(array(
                'groups' => array(
                    'value' => 'medics',
                )
            ),
                $this->datagridValues

            );
            return parent::getFilterParameters();
        }

    /**
     * @param DatagridMapper $datagridMapper
     */
    protected function configureDatagridFilters(DatagridMapper $datagridMapper)
    {


        $datagridMapper
            ->add('username')
            ->add('groups', 'doctrine_orm_callback', array(
                'callback' => array($this, 'queryGroupsByName'),
                'field_type' => 'text',
                'show_filter'=>false
            ))
            ->add('email')
            ->add('enabled')
            ->add('lastLogin', 'doctrine_orm_datetime_range', array('field_type' => 'sonata_type_datetime_range_picker'))
            //->add('lastLogin', 'doctrine_orm_datetime', array('field_type'=>'sonata_type_datetime_picker'))
            ->add('locked')
            ->add('roles')
            ->add('firstname')
            ->add('lastname')
            ->add('phoneno')
            ->add('effects')
            ->add('public_email')
            ->add('profilepic');
    }


    public function queryGroupsByName($queryBuilder, $alias, $field, $value)
    {
        if (empty($value['value'])) {
            return;
        }

        /** @var QueryBuilder $queryBuilder */
        $queryBuilder->leftJoin(sprintf('%s.groups', $alias), 'u');
        $queryBuilder->where('u.name LIKE :username');
        $queryBuilder->setParameter('username', '%' . $value['value'] . '%');
        return true;
    }

    /**
     * @param ListMapper $listMapper
     */
    protected function configureListFields(ListMapper $listMapper)
    {


        $listMapper
            ->add('username')
            ->add('email')
            ->add('locked', null, array('editable' => true))
            ->add('enabled')
            ->add('lastLogin')

            ->add('firstname')
            ->add('lastname')
            ->add('phoneno')
            ->add('profilepic', null, array(
                'template' => 'MedAppBundle:AdminSonata/CRUD:picture.html.twig'
            ))
            ->add('_action', 'actions', array(
                'actions' => array(
                    'show' => array(),
                    'edit' => array(),
                    'delete' => array(),
                )
            ));
    }

    /**
     * @param FormMapper $formMapper
     */
    protected function configureFormFields(FormMapper $formMapper)
    {
        $formMapper
            /*    ->tab('General')
                ->with('Addresses', array(
                    'class' => 'col-md-8',
                    'box_class' => 'box box-solid box-danger',
                    'description' => 'Lorem ipsum',
                ))*/
            ->add('username')
            ->add('usernameCanonical')
            ->add('email')
            ->add('emailCanonical')
            ->add('salt')
            ->add('password')
            ->add('lastLogin', 'sonata_type_datetime_picker')
            ->add('locked')
            ->add('expired')
            ->add('expiresAt')
            ->add('confirmationToken')
            ->add('passwordRequestedAt')
            ->add('roles')
            ->add('credentialsExpired')
            ->add('credentialsExpireAt')
            ->add('id')
            ->add('groups')
            ->add('firstname')
            ->add('lastname')
            ->add('phoneno')
            ->add('effects')
            ->add('public_email')
            ->add('profilepic', 'sonata_media_type', array(
                'provider' => 'sonata.media.provider.image',
                'context' => 'profile_pics'
            ))
            /*    ->add('images', 'sonata_type_collection', array(), array(
                    'edit' => 'inline',
                    'inline' => 'table',
                    'link_parameters' => array(
                        'context' => 'images',
                        'provider' => 'sonata.media.provider.image'
                    )
                ))*//*   ->end()
            ->end();*/
        ;
    }

    /**
     * @param ShowMapper $showMapper
     */
    protected function configureShowFields(ShowMapper $showMapper)
    {
        $showMapper
            ->add('username')
            ->add('usernameCanonical')
            ->add('email')
            ->add('emailCanonical')
            ->add('enabled')
            ->add('salt')
            ->add('password')
            ->add('lastLogin')
            ->add('locked')
            ->add('expired')
            ->add('expiresAt')
            ->add('confirmationToken')
            ->add('passwordRequestedAt')
            ->add('roles')
            ->add('credentialsExpired')
            ->add('credentialsExpireAt')
            ->add('id')
            ->add('groups')
            ->add('firstname')
            ->add('lastname')
            ->add('phoneno')
            ->add('effects')
            ->add('public_email')
            ->add('profilepic', null, array(
                'template' => 'MedAppBundle:AdminSonata/CRUD:picture.html.twig'
            ));
    }
    /*
        protected function configureRoutes(RouteCollection $collection)
        {
            $collection->add('clone', $this->getRouterIdParameter().'/clone');
        }*/
    /*  public function getTemplate($name)
      {
          switch ($name) {
              case 'edit':
                  return 'AcmeMyBundle::my-custom-edit.html.twig';
                  break;
              default:
                  return parent::getTemplate($name);
                  break;
          }
      }*/
}
```

```php
<?php

namespace Application\Sonata\AdminBundle\Admin;

use Doctrine\ORM\QueryBuilder;
use Sonata\AdminBundle\Admin\AbstractAdmin;
use Sonata\AdminBundle\Datagrid\DatagridMapper;
use Sonata\AdminBundle\Datagrid\ListMapper;
use Sonata\AdminBundle\Form\FormMapper;
use Sonata\AdminBundle\Route\RouteCollection;
use Sonata\AdminBundle\Show\ShowMapper;
use Symfony\Component\Form\Extension\Core\Type\TextType;

class PatientsAdmin extends AbstractAdmin
{

    protected $baseRoutePattern = 'patients';
    protected $baseRouteName = 'patients';
    protected $classnameLabel = 'Patients';
    
    
    protected $datagridValues = array(
        '_page' => 1,
        '_sort_order' => 'DESC',
        '_sort_by' => 'id'
    );



    public function getFilterParameters()
    {
        $this->datagridValues = array_merge(array(
            'groups' => array(
                'value' => 'patients',
            )
        ),
            $this->datagridValues

        );
        return parent::getFilterParameters();
    }

    /**
     * @param DatagridMapper $datagridMapper
     */
    protected function configureDatagridFilters(DatagridMapper $datagridMapper)
    {
        $datagridMapper
            ->add('username')
            ->add('groups', 'doctrine_orm_callback', array(
                'callback' => array($this, 'queryGroupsByName'),
                'field_type' => 'text',
                'show_filter'=>false
            ))
            ->add('email')
            ->add('enabled')
            ->add('lastLogin', 'doctrine_orm_datetime_range', array('field_type' => 'sonata_type_datetime_range_picker'))
            //->add('lastLogin', 'doctrine_orm_datetime', array('field_type'=>'sonata_type_datetime_picker'))
            ->add('locked')
            ->add('roles')
            ->add('firstname')
            ->add('lastname')
            ->add('phoneno')
            ->add('effects')
            ->add('public_email')
            ->add('profilepic');
    }

    public function queryGroupsByName($queryBuilder, $alias, $field, $value)
    {
        if (empty($value['value'])) {
            return;
        }

        /** @var QueryBuilder $queryBuilder */
        $queryBuilder->leftJoin(sprintf('%s.groups', $alias), 'u');
        $queryBuilder->where('u.name LIKE :username');
        $queryBuilder->setParameter('username', '%' . $value['value'] . '%');
        return true;
    }


    /**
     * @param ListMapper $listMapper
     */
    protected function configureListFields(ListMapper $listMapper)
    {
        $listMapper
            ->add('username')

            ->add('email')
            ->add('locked', null, array('editable' => true))
            ->add('enabled')


            ->add('lastLogin')
      

            ->add('firstname')
            ->add('lastname')
            ->add('phoneno')

            ->add('profilepic', null, array(
                'template' => 'MedAppBundle:AdminSonata/CRUD:picture.html.twig'
            ))
            ->add('_action', 'actions', array(
                'actions' => array(
                    'show' => array(),
                    'edit' => array(),
                    'delete' => array(),
                )
            ));
    }

    /**
     * @param FormMapper $formMapper
     */
    protected function configureFormFields(FormMapper $formMapper)
    {
        $formMapper
            /*    ->tab('General')
                ->with('Addresses', array(
                    'class' => 'col-md-8',
                    'box_class' => 'box box-solid box-danger',
                    'description' => 'Lorem ipsum',
                ))*/
            ->add('username')
            ->add('usernameCanonical')
            ->add('email')
            ->add('emailCanonical')
            ->add('salt')
            ->add('password')
            ->add('lastLogin', 'sonata_type_datetime_picker')
            ->add('locked')
            ->add('expired')
            ->add('expiresAt')
            ->add('confirmationToken')
            ->add('passwordRequestedAt')
            ->add('roles')
            ->add('credentialsExpired')
            ->add('credentialsExpireAt')
            ->add('id')
            ->add('groups')
            ->add('firstname')
            ->add('lastname')
            ->add('phoneno')
            ->add('effects')
            ->add('public_email')
            ->add('profilepic', 'sonata_media_type', array(
                'provider' => 'sonata.media.provider.image',
                'context' => 'profile_pics'
            ))
        /*    ->add('images', 'sonata_type_collection', array(), array(
                'edit' => 'inline',
                'inline' => 'table',
                'link_parameters' => array(
                    'context' => 'images',
                    'provider' => 'sonata.media.provider.image'
                )
            ))*//*   ->end()
            ->end();*/
        ;
    }

    /**
     * @param ShowMapper $showMapper
     */
    protected function configureShowFields(ShowMapper $showMapper)
    {
        $showMapper
            ->add('username')
            ->add('usernameCanonical')
            ->add('email')
            ->add('emailCanonical')
            ->add('enabled')
            ->add('salt')
            ->add('password')
            ->add('lastLogin')
            ->add('locked')
            ->add('expired')
            ->add('expiresAt')
            ->add('confirmationToken')
            ->add('passwordRequestedAt')
            ->add('roles')
            ->add('credentialsExpired')
            ->add('credentialsExpireAt')
            ->add('id')
            ->add('groups')
            ->add('firstname')
            ->add('lastname')
            ->add('phoneno')
            ->add('effects')
            ->add('public_email')
            ->add('profilepic', null, array(
                'template' => 'MedAppBundle:AdminSonata/CRUD:picture.html.twig'
            ));
    }
    /*
        protected function configureRoutes(RouteCollection $collection)
        {
            $collection->add('clone', $this->getRouterIdParameter().'/clone');
        }*/
    /*  public function getTemplate($name)
      {
          switch ($name) {
              case 'edit':
                  return 'AcmeMyBundle::my-custom-edit.html.twig';
                  break;
              default:
                  return parent::getTemplate($name);
                  break;
          }
      }*/
}


```

And a compiler pass

```php
class CompilerPass implements CompilerPassInterface
{
    public function process(ContainerBuilder $container)
    {
        $definition = $container->getDefinition('sonata.admin.pool');
        $definition->setClass('Application\Sonata\AdminBundle\Admin\Pool');
    }
}
```

**Comment by greg0ire** [comment]
OK, and when does the problem occur in your case ? Do you want to display the medic attached to a patient in the listing? Or what?

**Comment by clytemnestra** [comment]
The problem occurs when I access another admin that has a field which is in relation with my User table. For instance User-Picture,ManyToOne. The Picture admin will try to display user in the list, but it doesn't know which of the 2 admins defined for the User entity to use, so it throws the error.

**Comment by greg0ire** [comment]
> Since the entity is the same, it doesn't matter what admin is used.

If you're referencing the entity directly, then no, the admin does not matter at all indeed. So maybe the problem is that the TypeGuesser guesses an admin type, when it should guess an entity type ? Can you try using `add('userOrWhatever', 'entity')`, see if it works. If it does, then the type guesser probably has a bug.

**Comment by clytemnestra** [comment]
The main idea was to try and not modify every field that has add('user'), since there's a lot of them through the admin classes.

Anyways, I changed like you said in one of my classes and getting the same error.

**Comment by greg0ire** [comment]
> The main idea was to try and not modify every field that has add('user')

I know, but first, I would like to understand what should be done to automatically fill in the blanks. No you are getting the same error, which is very weird, since the admin should not be involved at all IMO if you use entity directly.

**Comment by greg0ire** [comment]
Could you post a stack trace, so that we can understand why `getAdminByClass` is called at all?

**Comment by clytemnestra** [comment]
Unable to find a valid admin for the class: MedAppBundle\Entity\User, there are too many registered: med_app.admin.user_medic,med_app.admin.user_pacient


```php

[1] RuntimeException: Unable to find a valid admin for the class: MedAppBundle\Entity\User, there are too many registered: med_app.admin.user_medic,med_app.admin.user_pacient
    at n/a
        in C:\htdocs\medapp\app\cache\dev\classes.php line 5611

    at Sonata\AdminBundle\Admin\Pool->getAdminByClass('MedAppBundle\Entity\User')
        in C:\htdocs\medapp\app\cache\dev\classes.php line 3936

    at Sonata\AdminBundle\Admin\AbstractAdmin->attachAdminClass(object(FieldDescription))
        in C:\htdocs\medapp\vendor\sonata-project\doctrine-orm-admin-bundle\Builder\ListBuilder.php line 139

    at Sonata\DoctrineORMAdminBundle\Builder\ListBuilder->fixFieldDescription(object(BannedUsersAdmin), object(FieldDescription))
        in C:\htdocs\medapp\vendor\sonata-project\doctrine-orm-admin-bundle\Builder\ListBuilder.php line 63

    at Sonata\DoctrineORMAdminBundle\Builder\ListBuilder->buildField('entity', object(FieldDescription), object(BannedUsersAdmin))
        in C:\htdocs\medapp\vendor\sonata-project\doctrine-orm-admin-bundle\Builder\ListBuilder.php line 71

    at Sonata\DoctrineORMAdminBundle\Builder\ListBuilder->addField(object(FieldDescriptionCollection), 'entity', object(FieldDescription), object(BannedUsersAdmin))
        in C:\htdocs\medapp\app\cache\dev\classes.php line 6294

    at Sonata\AdminBundle\Datagrid\ListMapper->add('user', 'entity')
        in C:\htdocs\medapp\src\Application\Sonata\AdminBundle\Admin\BannedUsersAdmin.php line 41

    at Application\Sonata\AdminBundle\Admin\BannedUsersAdmin->configureListFields(object(ListMapper))
        in C:\htdocs\medapp\app\cache\dev\classes.php line 4812

    at Sonata\AdminBundle\Admin\AbstractAdmin->buildList()
        in C:\htdocs\medapp\app\cache\dev\classes.php line 4189

    at Sonata\AdminBundle\Admin\AbstractAdmin->hasListFieldDescription('id')
        in C:\htdocs\medapp\app\cache\dev\classes.php line 3664

    at Sonata\AdminBundle\Admin\AbstractAdmin->buildDatagrid()
        in C:\htdocs\medapp\app\cache\dev\classes.php line 3974

    at Sonata\AdminBundle\Admin\AbstractAdmin->getDatagrid()
        in C:\htdocs\medapp\vendor\sonata-project\admin-bundle\Controller\CRUDController.php line 104

    at Sonata\AdminBundle\Controller\CRUDController->listAction()
        in  line 

    at call_user_func_array(array(object(BannedUsersAdminController), 'listAction'), array())
        in C:\htdocs\medapp\app\bootstrap.php.cache line 3211

    at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), '1')
        in C:\htdocs\medapp\app\bootstrap.php.cache line 3173

    at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), '1', true)
        in C:\htdocs\medapp\app\bootstrap.php.cache line 3324

    at Symfony\Component\HttpKernel\DependencyInjection\ContainerAwareHttpKernel->handle(object(Request), '1', true)
        in C:\htdocs\medapp\app\bootstrap.php.cache line 2514

    at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
        in C:\htdocs\medapp\web\app_dev.php line 28

```

**Comment by greg0ire** [comment]
I believe the offending piece of code is this one : https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/Builder/ListBuilder.php#L138

**Comment by greg0ire** [comment]
Can you try to comment it out?

**Comment by clytemnestra** [comment]
Commenting that line in both ListBuilder and DatagridBuilder will work, it will put the field as a string instead of link to the edit form.

**Comment by greg0ire** [comment]
> instead of a link

I don't get it. Isn't it supposed to be a select list rather than a link in the edit form ?

**Comment by clytemnestra** [comment]
The select list is for the datagrid aka filters, and the link is in the list field, in the table.

**Comment by greg0ire** [comment]
Oh ok I get it now. So this piece of code is maybe a bit too brutal, and should be fine-tuned to be called only when really needed.

**Comment by greg0ire** [comment]
> Commenting that line in both ListBuilder and DatagridBuilder will work, it will put the field as a string instead of link to the edit form.

Yeah so in fact it the admin does matter if we want to have links…

**Comment by clytemnestra** [comment]
Unless it doesn't matter which admin you want and either will do. Since both admins are for the same entity.

**Comment by greg0ire** [comment]
> Unless it doesn't matter which admin you want and either will do. Since both admins are for the same entity.

Yeah but sonata can't choose for you… so either it takes the first one, or we can maybe have a default admin code per entity, which feels better for people who would like to have links to users as patients 99% of the time

**Comment by clytemnestra** [comment]
That sounds right, how do we do that?

**Comment by greg0ire** [comment]
I think we need a new service and interface with a `getDefaultAdminForClass()` method.

**Comment by greg0ire** [comment]
`DefaultAdminResolverInterface`, for instance

**Comment by greg0ire** [comment]
I'll create a new issue for that

**Comment by greg0ire** [comment]
See #3908

**Pull request closed without merging**